### ISSUE TITLE
Worse performance using Accelerate

### ISSUE THREAD

**Issue by miguelscarv** [huggingface/accelerate#2341]
### System Info

```Shell
- `Accelerate` version: 0.26.1
- Platform: Linux-5.14.21-150500.55.36-default-x86_64-with-glibc2.31
- Python version: 3.11.4
- Numpy version: 1.26.0
- PyTorch version (GPU?): 2.1.1+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 1007.51 GB
- GPU type: NVIDIA A100 80GB PCIe
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: NO
        - mixed_precision: no
        - use_cpu: False
        - debug: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 3
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I am trying to use Accelerate to to distribute my training across multiple GPUs and I noticed that the training loss wasn't decreasing as fast as it should so I ran a new `accelerate config` to create the most basic one, without distributed training, to debug.

I seem to still be facing  the same problem, where the training loss decreases much slower, compared to a version of my script using only Pytorch. Here is the first version of training script I decided to use with Accelerate:
```
for epoch in range(1):

    model.train()
    dropout_modules = [module for module in model.lang_encoder.modules() if isinstance(module,torch.nn.Dropout)]
    [module.eval() for module in dropout_modules] # disable dropout
    accelerator.print(f"Disabled {len(dropout_modules)} Dropout modules")

    for i, batch in enumerate(tqdm(llavar_dataloader, disable=not accelerator.is_local_main_process)):

        with accelerator.accumulate(model):

            outputs = model(vision_x=batch["vision_x"], lang_x=batch["lang_x"].to(device), labels=batch["labels"].to(device), device=device)
            loss = outputs.loss
            if accelerator.is_local_main_process:
                total_loss += float(loss.item())
            accelerator.backward(loss)

            if accelerator.sync_gradients:
                accelerator.clip_grad_norm_(model.parameters(), 1.0)

            optimizer.step()
            model.zero_grad()

            if ((i+1) % accelerator.gradient_accumulation_steps == 0 or i+1 == len(llavar_dataloader)) and accelerator.is_local_main_process:
                with open("llavar_"+args.file, "a") as f:
                    f.write(f"Epoch: {epoch}, Loss: {total_loss/args.acc}\n")
                total_loss = 0 # Reset accumulated loss
```
**Here is the problem:** the training loss of both versions, using Accelerate and Pytorch, in the first iteration is the exact same. After that the loss decreases much faster in the pure Pytorch script. Note that I am using seeds in both, so that is why the first iteration is the exact same...

To debug I decided to remove the `.accumulate` call and build my own gradient accumulation flow, like this: 
``` 
for epoch in range(1):

    model.train()
    dropout_modules = [module for module in model.lang_encoder.modules() if isinstance(module,torch.nn.Dropout)]
    [module.eval() for module in dropout_modules] # disable dropout
    accelerator.print(f"Disabled {len(dropout_modules)} Dropout modules")

    for i, batch in enumerate(tqdm(llavar_dataloader, disable=not accelerator.is_local_main_process)):


        outputs = model(vision_x=batch["vision_x"], lang_x=batch["lang_x"].to(device), labels=batch["labels"].to(device), device=device)
        loss = outputs.loss
        if accelerator.is_local_main_process:
            total_loss += float(loss.item())
        accelerator.backward(loss / args.acc)            

        

        if ((i+1) % accelerator.gradient_accumulation_steps == 0 or i+1 == len(llavar_dataloader)) and accelerator.is_local_main_process:
            with open("llavar_"+args.file, "a") as f:
                f.write(f"Epoch: {epoch}, Loss: {total_loss/args.acc}\n")
            total_loss = 0 # Reset accumulated loss

            accelerator.clip_grad_norm_(model.parameters(), 1.0)

            optimizer.step()
            model.zero_grad()
``` 
and the results were very similar to the results in my Pytorch script when I removed gradient clipping. I also tested commenting out the line `accelerator.clip_grad_norm_(model.parameters(), 1.0)` and the results were the exact same. Note that added gradient clipping in my Pytorch script changes results a bit, and I also add it right before I call `optimizer.step()`.

It looks like this issue has something to do with how gradients are accumulated/clipped, but I can't seem to pinpoint the problem...

UPDATE: switching `accelerator.clip_grad_norm_` for the regular `torch.nn.utils.clip_grad_norm_` does nothing to the training loss - using any form of gradient clipping (or none) seems to yield the exact same results using Accelerate, which is the same training loss as my pure Pytorch script without gradient clipping, but adding gradient clipping in my Pytorch script further improves the training loss...

Also, here is my pure Pytorch training loop/script:
``` 
for epoch in range(1):

    model.train()
    dropout_modules = [module for module in model.lang_encoder.modules() if isinstance(module,torch.nn.Dropout)]
    [module.eval() for module in dropout_modules] # disable dropout
    print(f"Disabled {len(dropout_modules)} Dropout modules")

    for i, batch in enumerate(tqdm(coco_dataloader)):
        # Assume batch['image'] and batch['text'] are image and text tensors respectively
        outputs = model(vision_x=batch["vision_x"], lang_x=batch["lang_x"].to(f"cuda:{args.cuda}"), labels=batch["labels"].to(f"cuda:{args.cuda}"), device=f"cuda:{args.cuda}")
        loss = outputs.loss / num_accumulation_steps
        loss.backward()

        if (i+1) % num_accumulation_steps != 0:
            total_loss = loss.item() + total_loss
        # done once every num_accumulation_steps (5) steps
        elif (i+1) % num_accumulation_steps == 0:
            # Normalize loss
            total_loss = (total_loss + loss.item())
            # parameters updated
            torch.nn.utils.clip_grad_norm_(model.parameters(), 1)
            optimizer.step()
            # Reset gradient tensors
            model.zero_grad()
        
            with open("llavar_"+args.file, "a") as f:
                f.write(f"Epoch: {epoch}, Loss: {total_loss}\n")

            total_loss=0
```

### Expected behavior

I expected that using the Accelerate gradient accumulation method together with accelerate gradient clipping method should yield similar results to my pure Pytorch script, but something seems to be going wrong and it looks like it has to do with the gradients themselves...

**Comment by miguelscarv** [comment]
It looks like the error was `model.zero_grad()`. It is valid in vanilla Pytorch, but switching to  `optimizer.zero_grad()` achieves what I wanted. Still, I am not sure why grad clipping isn't working when I don't use the `.accumulate` context manager...
